Immutable/Apex.Analyzers.Immutable/Rules/Helper.cs

```diff
@@ -272,12 +272,17 @@ private static bool IsWhitelistedType(ITypeSymbol type, Compilation compilation)
                     break;
             }
 
-            if(
-                Equals(GetGuidType(compilation), type)
-                || Equals(GetTimeSpanType(compilation), type)
-                || Equals(GetDateTimeOffsetType(compilation), type)
-                || Equals(GetUriType(compilation), type)
-                )
+            if(type.TypeKind == TypeKind.Delegate)
+            {
+                return true;
```

Comment:
The problem is delegates aren't deeply immutable, which is what this analyzer is meant to enforce.  For example:
`
            int i = 0;
            Func<int> f = () => i += 1;
`
Every call f() will return a different value since the closure class is mutable.  I don't think there's any way to tell from static analysis even whether there are captured values or not, let alone their types.

I'm not sure what the right way to deal with delegates is, but I don't want to just allow all of them.

A consumer-side workaround would be to make wrapper structs for delegates and declare that struct with [Immutable(onFaith: true)] to skip the analysis for the delegate field.  For most use cases it wouldn't be a terrible amount of boilerplate since you just need one wrapper definition per delegate with differing number of args e.g., a wrapper for Func<T1>, Func<T1, T2>, etc.

Reply:
I started down the path of creating structs, but the loss of type inference makes the usage pretty ugly.

I created a separate PR for making the whitelist configurable. That will allow us to whitelist Action/Func without impacting other clients of the analyzers. We can live with the corner case of someone adding state through a closure. I'll close this PR.